### PR TITLE
OSDOCS#7562: Scheduling workloads on multi-arch clusters

### PR DESCRIPTION
**Pull request by kelbrown20**
**Version:** 4.14+ 

**Issue:** [OSDOCS-7562](https://issues.redhat.com//browse/OSDOCS-7562)

**Link to docs preview:**

[Managing your cluster with multi-architecture compute machines -> Scheduling workloads on clusters with multi-architecture compute machines 
](https://63915--docspreview.netlify.app/openshift-enterprise/latest/post_installation_configuration/configuring-multi-arch-compute-machines/multi-architecture-compute-managing)
**QE review:**
- [x] QE has approved this change.

### CODE REVIEW

File: modules/multi-architecture-scheduling-workloads.adoc [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+.Procedure 
+. Create a `RuntimeClass` object with the following command:
++
+-- 
+[source,terminal]
+----
+$ oc create -f <file-name>.yaml
+----
+.Example `RuntimeClass` object:
+[source,yaml]
```
Comment (kelbrown20) [inline modules/multi-architecture-scheduling-workloads.adoc:26]: @Prashanth684 I have a WIP here for scheduling workloads on multi-arch, the main part I', not sure about is what the RuntimeClass config should look like for a multi-arch workload.

File: modules/multi-architecture-scheduling-workloads.adoc [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+----
+apiVersion: node.k8s.io/v1beta1
+kind: RuntimeClass
+metadata:
+  name: <runtime_class_name> <1>
+handler: '' //not sure what these values are 
+scheduling:
+  nodeSelector: <2>
```
Comment (Prashanth684) [inline modules/multi-architecture-scheduling-workloads.adoc:34]: maybe a short description of the fact that we need nodeSelector's to ensure that the pod using the runtimeclass lands on the nodes that support it.

File: modules/multi-architecture-scheduling-workloads.adoc [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    kubernetes.io/os: 'linux'
+    kubernetes.io/arch: 'arm64'
+  tolerations:
```
Comment (Prashanth684) [inline modules/multi-architecture-scheduling-workloads.adoc:37]: another description of the case where we need tolerations - if the node that we want the pod to land on is tainted, we add a toleration. Instead of adding it to each pod we want, we just add these constraints to the runtimeclass and just make the pods use that for easy scheduling.

**Comment by Prashanth684** [comment]
@kelbrown20 can we organize this doc to be more in line with https://docs.openshift.com/container-platform/4.13/windows_containers/scheduling-windows-workloads.html ? the document should be about different ways workloads can be scheduled on a multi-arch compute cluster. i see you have only runtimeclass here, but it needs to be generic.

**Comment by kelbrown20** [comment]
Gotcha yeah! So add a more general description similar to this section in that doc (https://docs.openshift.com/container-platform/4.13/windows_containers/scheduling-windows-workloads.html#windows-pod-placement_scheduling-windows-workloads)?

**Comment by kelbrown20** [comment]
Also, do we still want the runtime class example here too?

File: modules/multi-architecture-scheduling-workloads.adoc (hunk 3 above)
Comment (kelbrown20) [reply modules/multi-architecture-scheduling-workloads.adoc:37]: Gotcha! Yeah that makes sense, and I can add a description like that as well

File: modules/multi-architecture-scheduling-examples.adoc [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+:_content-type: CONCEPT
+[id="multi-architecture-scheduling-examples_{context}"]
+
+= Sample multi-architecture node workload deployments 
+
+There are various cases to consider when you are scheduling workloads on a cluster with compute nodes of different architectures.
+
+//not sure about this format yet
+For example, you can taint a node using a `RuntimeClass` object to avoid unverified workloads, or workloads that are not compatible with an architecture, to be scheduled on that node.
```
Comment (Prashanth684) [inline modules/multi-architecture-scheduling-examples.adoc:13]: maybe we list some of these options as bullet points in more detail?

Comment (kelbrown20) [reply modules/multi-architecture-scheduling-examples.adoc:13]: Yeah, I was thinking that as well! I'll add some more of those cases we mentioned. Are there any example yaml that we have for those cases?

File: modules/multi-architecture-scheduling-examples.adoc [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+            nodeSelectorTerms:
+            - matchExpressions:
+              - key: kubernetes.io/arch
+                operator: In
+                values: <1>
+                - amd64
+                - arm64
+----
+<1> Specify the set of supported architecture. Valid values include `amd64` and `arm64`.
+//can this be set to the IBM arches as well?
```
Comment (Prashanth684) [inline modules/multi-architecture-scheduling-examples.adoc:37]: In theory, yes. But until we declare official support for P/Z we don't want to include those arches here

File: modules/multi-architecture-scheduling-examples.adoc [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+  template:
+    [...]
+    spec:
+      tolerations:
+      - key: "multi-arch.openshift.io/arch"
+        value: "arm64"
+        operator: "Equal"
+        effect: "NoSchedule"
+----
+//maybe I can add why a user would want to do this?
```
Comment (Prashanth684) [inline modules/multi-architecture-scheduling-examples.adoc:93]: Possibly: one reason could be that existing workload images might not be compatible with the architecture of the newly added node.

File: modules/multi-architecture-scheduling-examples.adoc [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
++
+[source,terminal]
+----
+$ oc annotate namespace my-namespace \
+  'scheduler.alpha.kubernetes.io/defaultTolerations'='[{"operator": "Exists", "effect": "NoSchedule", "key": "multi-arch.openshift.io/arch"}]'
+----
+
+Setting a toleration on a tainted node:: On a node with a defined taint, workloads will not be scheduled on that node. However, you can allow them to be scheduled by setting a toleration in a deployment 
+//what kind of Deployment is this yaml
++
```
Comment (Prashanth684) [inline modules/multi-architecture-scheduling-examples.adoc:75]: it could be any deployment. that doesn't matter.

File: post_installation_configuration/configuring-multi-arch-compute-machines/multi-architecture-compute-managing.adoc [hunk 8]
```diff
@@ -2,7 +2,24 @@
...
 [id="multi-architecture-compute-managing"]
 = Managing your cluster with multi-architecture compute machines
+include::_attributes/common-attributes.adoc[]
 
-toc::[] 
+toc::[]
+
+== Scheduling workloads on clusters with multi-architecture compute machines
+
+Before deploying a workload onto a cluster with compute nodes of different architectures, you must configure your compute node scheduling process so the pods in your cluster are correctly assigned.
```
Comment (aleskandro) [inline post_installation_configuration/configuring-multi-arch-compute-machines/multi-architecture-compute-managing.adoc:11]:
```suggestion
Deploying a workload onto a cluster with compute nodes of different architectures could require further attention to allow the successful placement of pods in the nodes of your cluster
```

File: modules/multi-architecture-scheduling-examples.adoc [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+//post_installation_configuration/configuring-multi-arch-compute-machines/multi-architecture-compute-managing.adoc
+
+:_content-type: CONCEPT
+[id="multi-architecture-scheduling-examples_{context}"]
+
+= Sample multi-architecture node workload deployments 
+
+There are various cases to consider when you are scheduling workloads on a cluster with compute nodes of different architectures. 
+
+Using node affinity to schedule workloads on a node:: To allow a workload to be scheduled on a certain architecture supported by its images, you can set the `spec.affinity.nodeAffinity` field in your pod specification deployment.
```
Comment (aleskandro) [inline modules/multi-architecture-scheduling-examples.adoc:12]:
```suggestion
Using node affinity to schedule workloads on a node:: To allow a workload to be scheduled only on a  set of nodes with architectures supported by its images, you can set the `spec.affinity.nodeAffinity` field in your pod template specification.
```

File: modules/multi-architecture-scheduling-examples.adoc [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+          requiredDuringSchedulingIgnoredDuringExecution:
+            nodeSelectorTerms:
+            - matchExpressions:
+              - key: kubernetes.io/arch
+                operator: In
+                values: <1>
+                - amd64
+                - arm64
+----
+<1> Specify the set of supported architecture. Valid values include `amd64` and `arm64`.
```
Comment (aleskandro) [inline modules/multi-architecture-scheduling-examples.adoc:36]: We can add `ppc64le` and `s390x` too.

File: modules/multi-architecture-scheduling-examples.adoc [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+
+Tainting nodes for a specific architecture:: You can taint a node to avoid workloads that are not compatible with an architecture to be scheduled on that node. In clusters using a `MachineSet` object, you can add parameters to a `.spec.template.spec.taints` field to avoid workloads being scheduled on nodes with non-supported architectures.
```
Comment (aleskandro) [inline modules/multi-architecture-scheduling-examples.adoc:38]:
```suggestion
Tainting nodes for a specific architecture:: You can taint a node to avoid workloads that are not compatible with its architecture to be scheduled on that node. In clusters using a `MachineSet` object, you can add parameters to a `.spec.template.spec.taints` field to avoid workloads being scheduled on nodes with non-supported architectures.
```

We should also say that users need to scale down the MachineSet or delete/rollout the already available Machines to updates the taints this way. This scale-in and out process is described in https://docs.openshift.com/container-platform/4.13/machine_management/modifying-machineset.html, that we could link here.

File: modules/multi-architecture-scheduling-examples.adoc [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+
+Creating a default toleration:: You can annotate a namespace so all of the workloads get the same default toleration by running the following command:
++
+[source,terminal]
+----
+$ oc annotate namespace my-namespace \
+  'scheduler.alpha.kubernetes.io/defaultTolerations'='[{"operator": "Exists", "effect": "NoSchedule", "key": "multi-arch.openshift.io/arch"}]'
+----
+
+Setting a toleration on a tainted node:: On a node with a defined taint, workloads will not be scheduled on that node. However, you can allow them to be scheduled by setting a toleration in a deployment.
```
Comment (aleskandro) [inline modules/multi-architecture-scheduling-examples.adoc:72]:
```suggestion
Tolerating the architecture taint in workloads:: On a node with a defined taint, workloads will not be scheduled on that node. However, you can allow them to be scheduled by setting a toleration in their pod template specification
```

File: modules/multi-architecture-scheduling-examples.adoc [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+  template:
+    [...]
+    spec:
+      tolerations:
+      - key: "multi-arch.openshift.io/arch"
+        value: "arm64"
+        operator: "Equal"
+        effect: "NoSchedule"
+----
+This deployment can be used when you have an existing workload image that is not compatible with the architecture of a new node.
```
Comment (aleskandro) [inline modules/multi-architecture-scheduling-examples.adoc:92]: The example deployment will be allowed on nodes with the `multi-arch.openshift.io/arch=arm64` taint

File: modules/multi-architecture-scheduling-examples.adoc [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+    spec:
+      tolerations:
+      - key: "multi-arch.openshift.io/arch"
+        value: "arm64"
+        operator: "Equal"
+        effect: "NoSchedule"
+----
+This deployment can be used when you have an existing workload image that is not compatible with the architecture of a new node. 
+
+Using node affinity with taints and tolerations:: When a scheduler assesses the node set, tolerations can broaden the set while node affinity restricts the set. The following example allows a workload to be scheduled on am `amd64` and `arm64` node with a toleration included.
```
Comment (aleskandro) [inline modules/multi-architecture-scheduling-examples.adoc:94]:
```suggestion
Using node affinity with taints and tolerations:: when the scheduler computes the set of candidate nodes to schedule a pod, tolerations can broaden the set while node affinity restricts the set. The following example allows a workload to be scheduled on `amd64` and `arm64` nodes. The toleration is required if you set a taint to the nodes of a given architecture, as depicted in the previous sections.
```

File: modules/multi-architecture-scheduling-runtimeclass.adoc [hunk 15]
```diff
@@ -0,0 +1,60 @@
+// module included in the following assembly
```
Comment (aleskandro) [inline modules/multi-architecture-scheduling-runtimeclass.adoc:1]: I would get rid of this since it's overkilling. It requires the users to set (and know about) the runtimeclass `handler`. It seems that runtimeclasses are meant specifically to change the handler and their scheduling fields are there for supporting this main feature.

Windows containers users must use this API, multi-arch ones don't need it

**Review by aleskandro — COMMENTED** [review]
Left some comments as suggestions to better clarify the scheduling context.

File: modules/multi-architecture-scheduling-examples.adoc (hunk 10 above)
Comment (kelbrown20) [reply modules/multi-architecture-scheduling-examples.adoc:36]: https://github.com/openshift/openshift-docs/pull/63915#discussion_r1316147010
Not sure about which ones

File: modules/multi-architecture-scheduling-examples.adoc (hunk 11 above)
Comment (Prashanth684) [reply modules/multi-architecture-scheduling-examples.adoc]: should we avoid these suggestions and just mention that users can taint nodes and leave it at that?

File: modules/multi-architecture-scheduling-runtimeclass.adoc (hunk 15 above)
Comment (Prashanth684) [reply modules/multi-architecture-scheduling-runtimeclass.adoc]: agreed. let's remove runtimeclass and its examples. It was relevant for  windows, but not here.

File: modules/multi-architecture-scheduling-examples.adoc [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+                operator: In
+                values: <1>
+                - amd64
+                - arm64
+----
+<1> Specify the set of supported architecture. Valid values include `amd64` and `arm64`.
+
+Tainting nodes for a specific architecture:: You can taint a node to avoid workloads that are not compatible with its architecture to be scheduled on that node. In clusters using a `MachineSet` object, you can add parameters to a `.spec.template.spec.taints` field to avoid workloads being scheduled on nodes with non-supported architectures.
+
+* Before you can taint a node, you must scale down the `MachineSet` object or remove available machines. You can scale down the machine set with the following commands:
```
Comment (kelbrown20) [inline modules/multi-architecture-scheduling-examples.adoc:40]: https://github.com/openshift/openshift-docs/pull/63915/files

I added some step and referenced the the scale-in and out process in the _Additional resources_ section. I think its a good process to mention if it is a requirement for the user. But I can also easily get rid of it. WDYT @aleskandro

Comment (aleskandro) [reply modules/multi-architecture-scheduling-examples.adoc:40]: sgtm!

File: modules/multi-architecture-scheduling-overview.adoc [hunk 17]
```diff
@@ -0,0 +1,13 @@
...
+//
+//post_installation_configuration/configuring-multi-arch-compute-machines/multi-architecture-compute-managing.adoc
+
+:_content-type: CONCEPT
+[id="multi-architecture-scheduling-overview_{context}"]
+= Scheduling workloads on clusters with multi-architecture compute machines 
+
+Before deploying a workload onto a cluster with compute nodes of different architectures, you must configure your compute node scheduling process so the pods in your cluster are correctly assigned. 
+
+There are various ways you can schedule workloads onto multi-architecture nodes for your cluster. You can use a node affinity, a node selector or a `RuntimeClass` object to select the node you want the pod to schedule onto. A `RuntimeClass` configuration also allows you to use scheduling mechanisms, like taints and tolderations.
```
Comment (aleskandro) [inline modules/multi-architecture-scheduling-overview.adoc:11]: should we remove the RuntimeClass bits here?

File: modules/multi-architecture-scheduling-examples.adoc [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+            - matchExpressions:
+              - key: kubernetes.io/arch
+                operator: In
+                values: <1>
+                - amd64
+                - arm64
+----
+<1> Specify the set of supported architecture. Valid values include `amd64` and `arm64`.
+
+Tainting nodes for a specific architecture:: You can taint a node to avoid workloads that are not compatible with its architecture to be scheduled on that node. In clusters using a `MachineSet` object, you can add parameters to a `.spec.template.spec.taints` field to avoid workloads being scheduled on nodes with non-supported architectures.
```
Comment (aleskandro) [inline modules/multi-architecture-scheduling-examples.adoc:38]: I'm not sure of the sentence 
```
In clusters using a `MachineSet` object
```

I used to mention it this way to exclude clusters like the ones installed via UPI methods because they do not leverage the MachineSet. However, I do not know how we refer to these cases in the doc for other scenarios.

Comment (jeffdyoung) [reply modules/multi-architecture-scheduling-examples.adoc:38]: Agree, lets reword to something like:

> Tainting every node for a specific architecture:: You can add a taint to a node to avoid workloads that are not compatible with its architecture.  In the case that your cluster can use a `MachineSet` object, you can add parameters to a ".spec.template.spec.taints" field to avoid workloads being scheduled on nodes with non-supported architectures.

File: post_installation_configuration/configuring-multi-arch-compute-machines/multi-architecture-compute-managing.adoc [hunk 19]
```diff
@@ -2,7 +2,26 @@
...
 [id="multi-architecture-compute-managing"]
 = Managing your cluster with multi-architecture compute machines
+include::_attributes/common-attributes.adoc[]
 
-toc::[] 
+toc::[]
+
+== Scheduling workloads on clusters with multi-architecture compute machines
+
+Deploying a workload onto a cluster with compute nodes of different architectures might require further attention in order to successfully place pods in the nodes of your cluster.
```
Comment (jeffdyoung) [inline post_installation_configuration/configuring-multi-arch-compute-machines/multi-architecture-compute-managing.adoc:11]: nit.. remove `might`

Comment (kelbrown20) [reply post_installation_configuration/configuring-multi-arch-compute-machines/multi-architecture-compute-managing.adoc:11]: Does deploying a workload always require further attention, or is this something users won't always have to do manually?

Comment (jeffdyoung) [reply post_installation_configuration/configuring-multi-arch-compute-machines/multi-architecture-compute-managing.adoc:11]: sorry..let me rephrase. Adding a 2nd arch to a cluster will always require attention to all new and existing workloads, but not necessarily action. Users have to make sure that everything deployed on their cluster can either run on both architectures, have the proper affinity and/or taints/tolerations, or some other means to guarantee proper scheduling. OCP and kubernetes will not do this for you.

Comment (kelbrown20) [reply post_installation_configuration/configuring-multi-arch-compute-machines/multi-architecture-compute-managing.adoc:11]: Ah I see what you're saying, hmm, what about something like this: 
> Deploying a workload onto a cluster with compute nodes of different architectures requires further attention, and might need further actions, in order to successfully place pods in the nodes of your cluster.

Comment (kelbrown20) [reply post_installation_configuration/configuring-multi-arch-compute-machines/multi-architecture-compute-managing.adoc:11]: @jeffdyoung After the discussion, wdyt of this: 
> Deploying a workload on a cluster with compute nodes of different architectures requires attention and monitoring of your cluster. There might be further actions you need to take in order to successfully place pods in the nodes of your cluster.

Comment (jeffdyoung) [reply post_installation_configuration/configuring-multi-arch-compute-machines/multi-architecture-compute-managing.adoc:11]: yeap. Thanks!

File: modules/multi-architecture-scheduling-examples.adoc [hunk 20]
```diff
@@ -0,0 +1,140 @@
...
+    spec:
+      tolerations:
+      - key: "multi-arch.openshift.io/arch"
+        value: "arm64"
+        operator: "Equal"
+        effect: "NoSchedule"
+----
+
+This deployment can be used when you have an existing workload image that is not compatible with the architecture of a new node. This example deployment will also be allowed on nodes with the `multi-arch.openshift.io/arch=arm64` taint specified.
+//not sure if both worked or if there was just meant to be one description here
```
Comment (aleskandro) [inline modules/multi-architecture-scheduling-examples.adoc:110]: Hey I think the first statement (*This deployment can be used when you have an existing workload image that is not compatible with the architecture of a new node*)  is out of the scope

Comment (kelbrown20) [reply modules/multi-architecture-scheduling-examples.adoc:110]: Whoops! I meant to remove that in the previous update. Removing now!

Comment (kelbrown20) [reply modules/multi-architecture-scheduling-examples.adoc]: @aleskandro Didn't you mention something about that sentence here, https://github.com/openshift/openshift-docs/pull/63915#discussion_r1321830944.

Comment (aleskandro) [reply modules/multi-architecture-scheduling-examples.adoc]: Hi @kelbrown20 @Prashanth684 The specific example we are discussing here adds toleration. I suggested the removal because, in this case, we allow the scheduling also on those nodes having the given taint, broadening the set of candidate nodes. Therefore, the user has (1) declared *to avoid* scheduling on the tainted nodes all the pods not tolerating for the taint, (2) added the toleration to a workload to declare that pod *supports/tolerate* the (tainted) architecture.

Comment (Prashanth684) [reply modules/multi-architecture-scheduling-examples.adoc]: ah..i see . By adding any sort of explanation we would probably be repeating what is already out there in the k8s docs, so makes sense to omit it.

**Comment by kelbrown20** [comment]
Is there anything else we need for this PR? Thank you sm for the reviews folks!

File: modules/multi-architecture-scheduling-examples.adoc [hunk 21]
```diff
@@ -0,0 +1,139 @@
...
+
+Creating a default toleration:: You can annotate a namespace so all of the workloads get the same default toleration by running the following command:
++
+[source,terminal]
+----
+$ oc annotate namespace my-namespace \
+  'scheduler.alpha.kubernetes.io/defaultTolerations'='[{"operator": "Exists", "effect": "NoSchedule", "key": "multi-arch.openshift.io/arch"}]'
+----
+
+Tolerating in workloads when the architecture taint is set:: On a node with a defined taint, workloads will not be scheduled on that node. However, you can allow them to be scheduled by setting a toleration in the pod specification.
```
Comment (aleskandro) [inline modules/multi-architecture-scheduling-examples.adoc:89]: > Tolerating in workloads when the architecture taint is set

this title seems missing an object: "tolerating architecture taints in workloads"?

File: modules/multi-architecture-scheduling-examples.adoc [hunk 22]
```diff
@@ -0,0 +1,139 @@
...
+      tolerations:
+      - key: "multi-arch.openshift.io/arch"
+        value: "arm64"
+        operator: "Equal"
+        effect: "NoSchedule"
+----
+
+This example deployment will also be allowed on nodes with the `multi-arch.openshift.io/arch=arm64` taint specified.
+
+Using node affinity with taints and tolerations:: When a scheduler computes the set of nodes to schedule a pod, tolerations can broaden the set while node affinity restricts the set. If you set a taint to the nodes of a specific architecture, the following example toleration is required for scheduling nodes.
```
Comment (aleskandro) [inline modules/multi-architecture-scheduling-examples.adoc:111]:
```suggestion
Using node affinity with taints and tolerations:: When a scheduler computes the set of nodes to schedule a pod, tolerations can broaden the set while node affinity restricts the set. If you set a taint to the nodes of a specific architecture, the following example toleration is required for scheduling pods.
```

**Review by aleskandro — COMMENTED** [review]
Other than the two comments below, this looks good to me

Tagging

/label qe-approved

**Comment by kelbrown20** [comment]
@Prashanth684 Does this also look good to you?

**Comment by Prashanth684** [comment]
> @Prashanth684 Does this also look good to you?

yes
/lgtm

File: modules/multi-architecture-scheduling-examples.adoc [hunk 23]
```diff
@@ -0,0 +1,139 @@
+// Module included in the following assembly 
+//
+//post_installation_configuration/configuring-multi-arch-compute-machines/multi-architecture-compute-managing.adoc
+
+:_content-type: CONCEPT
+[id="multi-architecture-scheduling-examples_{context}"]
+
+= Sample multi-architecture node workload deployments 
+
+There are various cases to consider when you are scheduling workloads on a cluster with compute nodes of different architectures.
```
Comment (dfitzmau) [inline modules/multi-architecture-scheduling-examples.adoc:10]: "There are" is a vague referent, so maybe something like:

> Before you schedule workloads on a cluster with compute nodes of different architectures, consider the following use cases:

File: modules/multi-architecture-scheduling-examples.adoc [hunk 24]
```diff
@@ -0,0 +1,139 @@
...
+//post_installation_configuration/configuring-multi-arch-compute-machines/multi-architecture-compute-managing.adoc
+
+:_content-type: CONCEPT
+[id="multi-architecture-scheduling-examples_{context}"]
+
+= Sample multi-architecture node workload deployments 
+
+There are various cases to consider when you are scheduling workloads on a cluster with compute nodes of different architectures. 
+
+Using node affinity to schedule workloads on a node:: To allow a workload to be scheduled on only a set of nodes with architectures supported by its images, you can set the `spec.affinity.nodeAffinity` field in your pod template specification.
```
Comment (dfitzmau) [inline modules/multi-architecture-scheduling-examples.adoc:12]:
```suggestion
Using node affinity to schedule workloads on a node:: To allow a workload to be scheduled on only a set of nodes with architectures supported by its images, you can set the `spec.affinity.nodeAffinity` field in your pod's template specification.
```

Comment (dfitzmau) [reply modules/multi-architecture-scheduling-examples.adoc:12]: "Allow" might be a form of anthropomorphism if it is directed at the user. But maybe in this case, the verb is directed at an object-to-object relationship? 

See: https://www.ibm.com/docs/en/ibm-style?topic=grammar-anthropomorphism

Comment (dfitzmau) [reply modules/multi-architecture-scheduling-examples.adoc:12]: "its" might get lost among the nouns (workload, nodes, architectures), maybe be more specific (that are supported byt the workload's images)?

File: modules/multi-architecture-scheduling-examples.adoc [hunk 25]
```diff
@@ -0,0 +1,139 @@
...
++
+.Example deployment with the `nodeAffinity` set to certain architectures 
+[source,yaml]
```
Comment (dfitzmau) [inline modules/multi-architecture-scheduling-examples.adoc:15]: See YAML guidance on https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#yaml-formatting-for-kubernetes-and-openshift-api-objects

Screenshot from 2023-09-25 17-09-16

File: modules/multi-architecture-scheduling-examples.adoc [hunk 26]
```diff
@@ -0,0 +1,139 @@
...
+          requiredDuringSchedulingIgnoredDuringExecution:
+            nodeSelectorTerms:
+            - matchExpressions:
+              - key: kubernetes.io/arch
+                operator: In
+                values: <1>
+                - amd64
+                - arm64
+----
+<1> Specify the set of supported architecture. Valid values include `amd64` and `arm64`.
```
Comment (dfitzmau) [inline modules/multi-architecture-scheduling-examples.adoc:36]: I was expecting more values with the word "set":

```suggestion

<1> Specify the supported architectures. Valid values include `amd64`,`arm64`, or both values.
```

File: modules/multi-architecture-scheduling-examples.adoc [hunk 27]
```diff
@@ -0,0 +1,139 @@
...
+
+Tainting every node for a specific architecture:: You can taint a node to avoid workloads that are not compatible with its architecture to be scheduled on that node. In the case that your cluster is using a `MachineSet` object, you can add parameters to a `.spec.template.spec.taints` field to avoid workloads being scheduled on nodes with non-supported architectures.
```
Comment (dfitzmau) [inline modules/multi-architecture-scheduling-examples.adoc:38]:
```suggestion
Tainting every node for a specific architecture:: You can taint a node to avoid workloads that are not compatible with its architecture to be scheduled on that node. In the case where your cluster is using a `MachineSet` object, you can add parameters to the `.spec.template.spec.taints` field to avoid workloads being scheduled on nodes with non-supported architectures.
```

File: modules/multi-architecture-scheduling-examples.adoc [hunk 28]
```diff
@@ -0,0 +1,139 @@
...
+
+* Before you can taint a node, you must scale down the `MachineSet` object or remove available machines. You can scale down the machine set with the following commands:
```
Comment (dfitzmau) [inline modules/multi-architecture-scheduling-examples.adoc:40]:
```suggestion
* Before you can taint a node, you must scale down the `MachineSet` object or remove available machines. You can scale down the machine set by using on of the following commands: 
```

File: modules/multi-architecture-scheduling-examples.adoc [hunk 29]
```diff
@@ -0,0 +1,139 @@
...
+        value: "arm64"
+        operator: "Equal"
+        effect: "NoSchedule"
+----
++
+This example deployment will also be allowed on nodes with the `multi-arch.openshift.io/arch=arm64` taint specified.
+
+Using node affinity with taints and tolerations:: When a scheduler computes the set of nodes to schedule a pod, tolerations can broaden the set while node affinity restricts the set. If you set a taint to the nodes of a specific architecture, the following example toleration is required for scheduling pods.
++
+.Example deployment with a node affinity and toleration set.
```
Comment (dfitzmau) [inline modules/multi-architecture-scheduling-examples.adoc:113]: Remove period for consistency with other example titles.

File: modules/multi-architecture-scheduling-examples.adoc [hunk 30]
```diff
@@ -0,0 +1,139 @@
...
+$ oc scale --replicas=0 machineset <machineset> -n openshift-machine-api
+----
++
+Or: 
++
+[source,terminal]
+----
+$ oc edit machineset <machineset> -n openshift-machine-api
+----
+For more information on scaling machine sets, see "Modifying a compute machine set".
```
Comment (dfitzmau) [inline modules/multi-architecture-scheduling-examples.adoc:53]:
```suggestion
For more information on scaling machine sets, see the "Modifying a compute machine set" section.
```

File: modules/multi-architecture-scheduling-examples.adoc [hunk 31]
```diff
@@ -0,0 +1,139 @@
...
+
```
Comment (dfitzmau) [inline modules/multi-architecture-scheduling-examples.adoc:54]: Remove line 54 so that the plus symbol aligns with the p[previous step. This should fix the indent issue.

File: modules/multi-architecture-scheduling-examples.adoc [hunk 32]
```diff
@@ -0,0 +1,139 @@
...
+  template:
+    [...]
+    spec:
+      [...]
+      taints:
+      - effect: NoSchedule         
+        key: multi-arch.openshift.io/arch                
+        value: arm64
+----
+You can also set a taint on a specific node by running the following command:
```
Comment (dfitzmau) [inline modules/multi-architecture-scheduling-examples.adoc:74]: Maybe numeric steps as this might be an optionals step 2?

File: modules/multi-architecture-scheduling-examples.adoc [hunk 33]
```diff
@@ -0,0 +1,139 @@
...
+    [...]
+    spec:
+      tolerations:
+      - key: "multi-arch.openshift.io/arch"
+        value: "arm64"
+        operator: "Equal"
+        effect: "NoSchedule"
+----
++
+This example deployment will also be allowed on nodes with the `multi-arch.openshift.io/arch=arm64` taint specified.
```
Comment (dfitzmau) [inline modules/multi-architecture-scheduling-examples.adoc:109]: Just  wondering if the future tense is needed:
Screenshot from 2023-09-25 17-17-54

Also, maybe more active sentence could help distinguish the agent? I'm unsure what is allowing the action here.

File: modules/multi-architecture-scheduling-examples.adoc [hunk 34]
```diff
@@ -0,0 +1,139 @@
...
+
+Creating a default toleration:: You can annotate a namespace so all of the workloads get the same default toleration by running the following command:
++
+[source,terminal]
+----
+$ oc annotate namespace my-namespace \
+  'scheduler.alpha.kubernetes.io/defaultTolerations'='[{"operator": "Exists", "effect": "NoSchedule", "key": "multi-arch.openshift.io/arch"}]'
+----
+
+Tolerating architecture taints in workloads:: On a node with a defined taint, workloads will not be scheduled on that node. However, you can allow them to be scheduled by setting a toleration in the pod specification.
```
Comment (dfitzmau) [inline modules/multi-architecture-scheduling-examples.adoc:89]:
```suggestion
Tolerating architecture taints in workloads:: On a node with a defined taint, workloads will not be scheduled on that node. However, you can allow them to be scheduled by setting a toleration in the pod's specification.
```

File: modules/multi-architecture-scheduling-overview.adoc [hunk 35]
```diff
@@ -0,0 +1,13 @@
+// module included in the following assembly 
+//
+//post_installation_configuration/configuring-multi-arch-compute-machines/multi-architecture-compute-managing.adoc
+
+:_content-type: CONCEPT
+[id="multi-architecture-scheduling-overview_{context}"]
+= Scheduling workloads on clusters with multi-architecture compute machines 
+
+Before deploying a workload onto a cluster with compute nodes of different architectures, you must configure your compute node scheduling process so the pods in your cluster are correctly assigned.
```
Comment (dfitzmau) [inline modules/multi-architecture-scheduling-overview.adoc:9]: Nice!

File: modules/multi-architecture-scheduling-overview.adoc [hunk 36]
```diff
@@ -0,0 +1,13 @@
...
+//
+//post_installation_configuration/configuring-multi-arch-compute-machines/multi-architecture-compute-managing.adoc
+
+:_content-type: CONCEPT
+[id="multi-architecture-scheduling-overview_{context}"]
+= Scheduling workloads on clusters with multi-architecture compute machines 
+
+Before deploying a workload onto a cluster with compute nodes of different architectures, you must configure your compute node scheduling process so the pods in your cluster are correctly assigned. 
+
+There are various ways you can schedule workloads onto multi-architecture nodes for your cluster. You can use a node affinity or a node selector to select the node you want the pod to schedule onto. You can also use scheduling mechanisms, like taints and tolderations, when using node affinity or node selector to correctly schedule workloads.
```
Comment (dfitzmau) [inline modules/multi-architecture-scheduling-overview.adoc:11]:
```suggestion
You can schedule workloads onto multi-architecture nodes for your cluster in serveral ways. For example, you can use a node affinity or a node selector to select the node you want the pod to schedule onto. You can also use scheduling mechanisms, like taints and tolderations, when using node affinity or node selector to correctly schedule workloads. 
```

**Review by dfitzmau — COMMENTED** [review]
Nice work, @kelbrown20 . I added comments inline.

/remove-label peer-review-in-progress

/label peer-review-done

File: modules/multi-architecture-scheduling-examples.adoc (hunk 24 above)
Comment (kelbrown20) [reply modules/multi-architecture-scheduling-examples.adoc:12]: Ah gotcha, yeah I see what you mean, lemme re word that. Thank you!

File: modules/multi-architecture-scheduling-examples.adoc (hunk 32 above)
Comment (kelbrown20) [reply modules/multi-architecture-scheduling-examples.adoc:74]: These are all kinda optional, its more for the user to decide on which way they want to schedule nodes. So its more of a reference
